### PR TITLE
Session Issue #3

### LINKED ISSUE #3: As a debugging user, I can enter a specific waiting room to call upon an available rubber duck.
Allow users to enter a size-limited voice channel to indicate their desire to begin a debugging session. If no members with a rubber duck role are available, this automation should restrict entry to this channel. Upon entry, this automation should:
- increase the user limit of the voice channel from one to two
- limit the further entry to the voice channel to users with the rubber duck role
- send a message to all available members with the rubber duck role, directing them to join the debugging user in the specified voice channel.

### PR DESCRIPTION
**Pull request by kobito-kun**
Issue #3

### CODE REVIEW

File: quackfactory/cogs/session.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+@client.event
+async def on_voice_state_update(member, before, after):
+
+    guild = member.guild
+    role = "Test"
+    everyone = "@everyone" 
+    channel_id = 756574459556659323
+    message = f"Hey! {member} needs a rubber duck!"
+
+    if before.channel is None and after.channel.id == channel_id:
```
Comment (amazingproducer) [inline quackfactory/cogs/session.py]: Is there a reason you chose to only react to users who were not in other voice channels?

File: quackfactory/cogs/session.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        await after.channel.set_permissions(discord.utils.get(guild.roles, name=role), connect=True)
+        await after.channel.set_permissions(discord.utils.get(guild.roles, name=everyone), connect=False)
+        await after.channel.edit(user_limit=2)
+
+        for member in guild.members:
+            for role in member.roles:
```
Comment (amazingproducer) [inline quackfactory/cogs/session.py]: Our specification dictates that we notify *available* ducks; please consider using presence intents to retrieve [user status](https://discordpy.readthedocs.io/en/latest/api.html#discord.Status).

**Review by amazingproducer — COMMENTED** [review]
This PR appears to solve #3, though it claims to solve #4. I've included some additional comments. Let's see if the rest of the team has any input.

File: quackfactory/cogs/session.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+from discord.ext import commands
+import discord
+import logging
+
+intents = discord.Intents.default()
+intents.reactions = True
+intents.members = True
+client = discord.Client(intents=intents)
```
Comment (hum) [inline quackfactory/cogs/session.py]: Why are you creating a separate instance of a bot inside of the cog? I recommend to go through the project and follow the code flow. Declaring intents and creating a new bot instance isn't what cogs are for -- these things are already created inside of the application.

PS: You are importing the logging package, yet you aren't using it.

**Review by hum — CHANGES_REQUESTED** [review]

**Comment by kobito-kun** [comment]
I'm so sorry :-;

File: quackfactory/cogs/session.py (hunk 3 above)
Comment (kobito-kun) [reply quackfactory/cogs/session.py]: i'll remove those 👍

**Comment by kobito-kun** [comment]
Hello, solved all the issues above. Please re-check. I apologise for not understanding well.

**Comment by amazingproducer** [comment]
Looks better! Let's see what @hum has to say.

File: quackfactory/cogs/session.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    if after.channel.id == channel_id:
+        await after.channel.set_permissions(discord.utils.get(guild.roles, name=role), connect=True)
+        await after.channel.set_permissions(discord.utils.get(guild.roles, name=everyone), connect=False)
+        await after.channel.edit(user_limit=2)
+
+        for member in guild.members:
+            for role in member.roles:
+                if role.name == "Test": # Rubberduck role idk why but it doesnt work using a variable 
+                    if str(member.status) == "online" or member.status == discord.Status.online:
```
Comment (hum) [inline quackfactory/cogs/session.py]: `if str(member.status) == "online" or member.status == discord.Status.online:`
These both evaluate the same thing.
```suggestion
                    if member.status == discord.Status.online:
```

**Review by hum — CHANGES_REQUESTED** [review]

File: quackfactory/cogs/session.py (hunk 4 above)
Comment (hum) [reply quackfactory/cogs/session.py]: You also need to add the proper intents in config.py for the status to work.

File: quackfactory/cogs/session.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import discord
+
+class Session(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+
+@client.event
```
Comment (hum) [inline quackfactory/cogs/session.py]:
```suggestion
@commands.Cog.listener()
```
You are using the wrong bot object for this to work as a cog.

it has to be the `discord.commands` bot instance. Please refer to [reactions.py](https://github.com/silobusters/quackfactory/blob/master/quackfactory/cogs/reactions.py)

**Review by hum — CHANGES_REQUESTED** [review]

File: quackfactory/cogs/session.py (hunk 4 above)
Comment (kobito-kun) [reply quackfactory/cogs/session.py]: Hi, I know this but I'll remove it.

File: quackfactory/cogs/session.py (hunk 5 above)
Comment (kobito-kun) [reply quackfactory/cogs/session.py]: Alright.

**Comment by kobito-kun** [comment]
Alright, added presences intents, used correct bot object and cleaned check member status.

File: quackfactory/cogs/session.py [hunk 6]
```diff
@@ -21,7 +22,7 @@ async def on_voice_state_update(member, before, after):
         for member in guild.members:
             for role in member.roles:
                 if role.name == "Test": # Rubberduck role idk why but it doesnt work using a variable
```
Comment (amazingproducer) [inline quackfactory/cogs/session.py]: Perhaps you could compare the [role id](https://discordpy.readthedocs.io/en/latest/api.html?highlight=role%20id#discord.Role.id) to the named variable for the role.

Comment (kobito-kun) [reply quackfactory/cogs/session.py]: Alright.

File: quackfactory/cogs/session.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+intents.reactions = True
+intents.members = True
+client = discord.Client(intents=intents)
+
+class Session(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+
+@client.event
+async def on_voice_state_update(member, before, after):
```
Comment (hum) [inline quackfactory/cogs/session.py]: Due to the nature of Python relying on indentation to recognise which methods belong to which class, you need to indent this to be inside of the session class.

**Review by hum — CHANGES_REQUESTED** [review]

**Comment by hum** [comment]
LGTM

**Pull request merged**